File: pt_seq_nms/seq_nms.py
from typing import List, Tuple

import torch


def _all_same_types(types: List[str]) -> bool:
    # Can't use set(types) == 1 since set() is not supported by torchscript at the moment
    found: List[str] = []

    for t in types:
        if t not in found:
            found.append(t)

    return len(found) == 1


def _validate_tensor_types(boxes: torch.Tensor, scores: torch.Tensor, classes: torch.Tensor) -> None:
    """
    Utility function for validating that the boxes, scores and classes have the same type.

    Args:
        boxes (Tensor[F, N, 4]) Boxes to perform seq-nms on. They are expected to be in
           (x_min, y_min, x_max, y_max) format.
        scores (Tensor[F, N]): Scores for each one of the boxes.
        classes (Tensor[F, N]): Class for each one of the boxes.
    Returns:
    """

    assert boxes.dtype == torch.float32, f"boxes are expected to have dtype float32, got {boxes.dtype}"
    assert scores.dtype == torch.float32, f"scores are expected to have dtype float32, got {scores.dtype}"
    assert classes.dtype == torch.int32, f"classes are expected to have dtype float32, got {classes.dtype}"
    assert _all_same_types(
        [boxes.device.type, scores.device.type, classes.device.type]
    ), "Expected all of the tensors to be on the same device"


def _validate_auxiliary_params(linkage_threshold: float, iou_threshold: float, metrics: str = "avg") -> None:
    """
    Utility function for validating that the parameters are in the valid range.

    Args:
        linkage_threshold (float): The threshold for linking two objects in consecutive frames.
        iou_threshold (float): the threshold for considering two boxes to be overlapping.
        metric (str): the metric type, currently "avg" and "max" is supported.
    Returns:
    """

    assert 0.0 <= linkage_threshold <= 1.0, f"linkage_threshold should be between 0.0 and 1.0; got {linkage_threshold}"
    assert 0.0 <= iou_threshold <= 1.0, f"iou_threshold should be between 0.0 and 1.0; got {iou_threshold}"
    assert metrics in ("avg", "max"), f"Expected metrics to be in {('avg', 'max')}, got {metrics}"


def seq_nms(
    boxes: torch.Tensor,
    scores: torch.Tensor,
    classes: torch.Tensor,
    linkage_threshold: float,
    iou_threshold: float,
    metrics: str = "avg",
) -> torch.Tensor:
    """
    Applies the seq-nms algorithm to the input boxes.

    Below F is the number of frames and N is the number of objects per frame.

    Args:
        boxes (Tensor[F, N, 4]) Boxes to perform seq-nms on. They are expected to be in
           (x_min, y_min, x_max, y_max) format.
        scores (Tensor[F, N]): Scores for each one of the boxes.
        classes (Tensor[F, N]): Class for each one of the boxes.
        linkage_threshold (float): The threshold for linking two objects in consecutive frames.
        iou_threshold (float): the threshold for considering two boxes to be overlapping.
        metric (str): the metric type, currently "avg" and "max" is supported.
    Returns:
        updated_scores (Tensor): tensor with the updated scores, i.e. the scores after they have been
            updated according to seq-nms.
    """

    assert len(boxes.shape) == 3 and boxes.shape[-1] == 4, f"boxes has wrong shape, expected (F, N, 4), got {boxes.shape}"
    assert len(scores.shape) == 2, f"scores has wrong shape, expected (F, N) got {scores.shape}"
    assert len(classes.shape) == 2, f"classes has wrong shape, expected (F, N) got {classes.shape}"

    _validate_tensor_types(boxes, scores, classes)
    _validate_auxiliary_params(linkage_threshold, iou_threshold, metrics)

    updated_scores = torch.ops.seq_nms.seq_nms(boxes, scores, classes, linkage_threshold, iou_threshold, metrics)
    return updated_scores


def _from_list_to_tensor(
    boxes_list: List[torch.Tensor],
    scores_list: List[torch.Tensor],
    classes_list: List[torch.Tensor],
) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    """
    Converts a list of tensors of different shapes to one tensor by taking the max-length
        and filling the remaining values with zeros.

    For example, if the input is [[1], [2, 3]] the output will be [[1, 0], [2, 3]].

    Below F is the number of frames and N_f is the number of objects per frame (which can vary between frames).
    M is the maximum length of each box.

    Args:
        boxes_list (List[F]) Boxes to perform seq-nms on. Each element in the list is a Tensor[N_f, 4].
            They are expected to be in (x_min, y_min, x_max, y_max) format.
        scores_list (List[F]) Scores for each one of the boxes. Each element in the list is a Tensor[N_f].
        classes_list (List[F]): Class for each one of the boxes. Each element in the list is a Tensor[N_f].
    Returns:
        boxes (Tensor[F, M, 4]): Concatenated boxes.
        scores (Tensor[F, M]): Concatenated scores.
        classes (Tensor[F, M]): Concatenated classes.
    """

    max_length = max([len(b) for b in boxes_list])
    num_element = len(boxes_list)
    device = boxes_list[0].device.type if len(boxes_list) > 0 else "cpu"

    boxes = torch.zeros((num_element, max_length, 4), dtype=torch.float32, device=device)
    scores = torch.zeros((num_element, max_length), dtype=torch.float32, device=device)
    classes = -1 * torch.ones((num_element, max_length), dtype=torch.int32, device=device)

    for idx in range(len(boxes_list)):
        len_idx = len(boxes_list[idx])

        boxes_idx = boxes_list[idx]
        scores_idx = scores_list[idx]
        classes_idx = classes_list[idx]

        _validate_tensor_types(boxes_idx, scores_idx, classes_idx)

        boxes[idx, 0:len_idx, :] = boxes_idx
        scores[idx, 0:len_idx] = scores_idx
        classes[idx, 0:len_idx] = classes_idx

    return boxes, scores, classes


def seq_nms_from_list(
    boxes_list: List[torch.Tensor],
    scores_list: List[torch.Tensor],
    classes_list: List[torch.Tensor],
    linkage_threshold: float,
    iou_threshold: float,
    metrics: str = "avg",
) -> torch.Tensor:
    """
    Applies the seq-nms algorithm to a list of input boxes, which can have different shapes.

    Below F is the number of frames and N_f is the number of objects per frame (which can vary between frames).

    Args:
        boxes (List[F]) Boxes to perform seq-nms on. Each element in the list is a Tensor[N_f, 4].
            They are expected to be in (x_min, y_min, x_max, y_max) format.
        scores (List[F]) Scores for each one of the boxes. Each element in the list is a Tensor[N_f].
        classes (List[F]): Class for each one of the boxes. Each element in the list is a Tensor[N_f].
        linkage_threshold (float): The threshold for linking two objects in consecutive frames.
        iou_threshold (float): the threshold for considering two boxes to be overlapping.
        metric (str): the metric type, currently "avg" and "max" is supported.
    Returns:
        updated_scores (Tensor): tensor with the updated scores, i.e. the scores after they have been
            updated according to seq-nms.
    """

    _validate_auxiliary_params(linkage_threshold, iou_threshold, metrics)

    boxes, scores, classes = _from_list_to_tensor(boxes_list, scores_list, classes_list)

    updated_scores = torch.ops.seq_nms.seq_nms(boxes, scores, classes, linkage_threshold, iou_threshold, metrics)
    return updated_scores

File: pt_seq_nms/csrc/seq_nms.cpp
#include "seq_nms.h"
#include <exception>
#include <tuple>
#include <vector>
#include "box_utils.h"
#include "sequence_utils.h"

using namespace torch::indexing;

box_seq_t build_box_sequences(
    const torch::Tensor& boxes,
    const torch::Tensor& box_areas,
    const torch::Tensor& classes,
    const double& linkage_threshold) {
    /*
    Creates a graph where vericies are object at a given frame and the edges are if they have an IOU higher than
    @linkage_threshold (two consecutive frames).

    boxes are expected to have the shape [F, N, 4] and of the format [x_min, y_min, x_max, y_max].
        F is the number of frames and N is the number of objects per frame.
    classes are expected to have the shape [F, N].
    linkage_threshold is the threshold for linking two objects in consecutive frames.
    */

    auto classes_acc = classes.accessor<int, 2>();

    box_seq_t box_graph;
    for (int f_idx = 0; f_idx < boxes.size(0) - 1; f_idx++) {
        auto current_box = boxes.index({f_idx, Slice(), Slice()});
        auto current_area = box_areas.index({f_idx, Slice()});

        auto next_box = boxes.index({f_idx + 1, Slice(), Slice()});
        auto next_area = box_areas.index({f_idx + 1, Slice()});

        // overlap has shape [N, N]
        auto overlaps = calculate_iou_given_area(current_box, next_box, current_area, next_area);
        auto overlaps_acc = overlaps.accessor<float, 2>();

        std::vector<std::vector<int>> adjacency_matrix;
        for (int b_idx = 0; b_idx < current_box.size(0); b_idx++) {
            std::vector<int> edges;
            for (int ovr_idx = 0; ovr_idx < overlaps.size(1); ovr_idx++) {
                float iou = overlaps_acc[b_idx][ovr_idx];
                bool same_class = (classes_acc[f_idx][b_idx] == classes_acc[f_idx + 1][ovr_idx]);

                // class idx < 0 are considered skip idxs
                bool accepted_class = (classes_acc[f_idx][b_idx] >= 0);

                if ((iou >= linkage_threshold) && same_class && accepted_class) {
                    edges.push_back(ovr_idx);
                }
            }

            adjacency_matrix.push_back(edges);
        }
        box_graph.push_back(adjacency_matrix);
    }

    return box_graph;
}

ScoreMetric get_score_enum_from_string(const std::string& metric_string) {
    /*
    Converts @metric_string to the enum "ScoreMetric".
    */

    if (metric_string == "avg") {
        return ScoreMetric::avg;
    } else if (metric_string == "max") {
        return ScoreMetric::max;
    } else {
        throw std::invalid_argument("Unsupported metric_string");
    }
}

torch::Tensor seq_nms(
    const torch::Tensor& boxes,
    const torch::Tensor& scores,
    const torch::Tensor& classes,
    const double& linkage_threshold,
    const double& iou_threshold,
    const std::string& metric) {
    /*
    Applies the seq-nms algorithm to the input boxes.

    boxes are expected to have the shape [F, N, 4] and of the format [x_min, y_min, x_max, y_max].
        F is the number of frames and N is the number of objects per frame.
    scores are expected to have the shape [F, N].
    classes are expected to have the shape [F, N].
    linkage_threshold is the threshold for linking two objects in consecutive frames.
    iou_threshold is the threshold for considering two boxes to be overlapping.
    metric is the metric type, currently "avg" and "max" is supported.
    */

    ScoreMetric metric_enum = get_score_enum_from_string(metric);
    float linkage_threshold_float = static_cast<float>(linkage_threshold);
    float iou_threshold_float = static_cast<float>(iou_threshold);

    torch::Tensor box_areas = calculate_area(boxes);
    box_areas = box_areas.to(torch::kCPU);

    const auto boxes_cpu = boxes.to(torch::kCPU);
    const auto classes_cpu = classes.to(torch::kCPU);

    box_seq_t box_graph = build_box_sequences(boxes_cpu, box_areas, classes_cpu, linkage_threshold_float);
    torch::Tensor local_scores = scores.to(torch::kCPU).clone();

    while (true) {
        auto best_tuple = find_best_sequence(box_graph, local_scores);

        int sequence_frame_index = std::get<0>(best_tuple);
        std::vector<int> best_sequence = std::get<1>(best_tuple);
        float best_score = std::get<2>(best_tuple);

        if (best_sequence.size() <= 1) {
            break;
        }

        rescore_sequence(best_sequence, local_scores, sequence_frame_index, best_score, metric_enum);
        delete_sequence(
            best_sequence, sequence_frame_index, local_scores, boxes_cpu, box_areas, box_graph, iou_threshold_float);
    }

    local_scores = local_scores.to(scores.device());
    return local_scores;
}
